test/unit/math/rev/mat/fun/matrix_exp_action_test.cpp

```diff
@@ -0,0 +1,368 @@
+  for (int l = 0; l < res_vd.size(); ++l) {
+    EXPECT_FLOAT_EQ(res_vd(l).val(), expAB(l).val());
+  }
+
+  // compare adjoints
+  std::vector<double> g, g0;
+  for (int l = 0; l < M; ++l) {
+    for (int k = 0; k < N; ++k) {
+      stan::math::set_zero_all_adjoints();
+      res_vd(k, l).grad(Avec, g);
+      stan::math::set_zero_all_adjoints();
+      expAB(k, l).grad(Avec, g0);
+      for (size_t j = 0; j < g.size(); ++j) {
+        EXPECT_FLOAT_EQ(g[j], g0[j]);
+      }
+    }
+  }
+
+  // test a single function of expA*B
+  var f = sum(res_vd);
+  var f0 = sum(expAB);
+  stan::math::set_zero_all_adjoints();
+  f.grad(Avec, g);
+  stan::math::set_zero_all_adjoints();
+  f0.grad(Avec, g0);
+  for (size_t j = 0; j < g.size(); ++j) {
+    EXPECT_FLOAT_EQ(g[j], g0[j]);
+  }
+}
+
+TEST(MathMatrix, matrix_exp_action_vd) {
+  test_matrix_exp_action_vd<1, 1>();
+  test_matrix_exp_action_vd<1, 5>();
+  test_matrix_exp_action_vd<5, 1>();
+  test_matrix_exp_action_vd<5, 5>();
+}
+
+template <int N, int M>
+inline void test_matrix_exp_action_vv() {
+  using stan::math::value_of;
+  using stan::math::var;
+  stan::math::matrix_exp_action_handler handler;
+  std::srand(2999);
+
+  Eigen::Matrix<var, N, N> Av = Eigen::Matrix<var, N, N>::Random();
+  Eigen::Matrix<var, N, M> Bv = Eigen::Matrix<var, N, M>::Random();
+  std::vector<stan::math::var> Bvec(Bv.data(), Bv.data() + Bv.size());
+  std::vector<stan::math::var> Avec(Av.data(), Av.data() + Av.size());
+
+  // brute force
+  Eigen::Matrix<var, N, N> expA = stan::math::matrix_exp(Av);
+  Eigen::Matrix<var, N, M> expAB;
+  for (int i = 0; i < N; ++i) {
+    for (int j = 0; j < M; ++j) {
+      expAB(i, j) = 0.0;
+      for (int k = 0; k < N; ++k) {
+        expAB(i, j) += expA(i, k) * Bv(k, j);
+      }
+    }
+  }
+
+  // matrix_exp_action
+  Eigen::Matrix<var, N, M> res_vv = stan::math::matrix_exp_action(Av, Bv);
+  for (int l = 0; l < res_vv.size(); ++l) {
+    EXPECT_FLOAT_EQ(res_vv(l).val(), expAB(l).val());
+  }
+  Avec.insert(Avec.end(), Bvec.begin(), Bvec.end());
+  // compare adjoints
+  std::vector<double> g, g0;
+  for (int l = 0; l < M; ++l) {
+    for (int k = 0; k < N; ++k) {
+      stan::math::set_zero_all_adjoints();
+      res_vv(k, l).grad(Avec, g);
+      stan::math::set_zero_all_adjoints();
+      expAB(k, l).grad(Avec, g0);
+      for (size_t j = 0; j < g.size(); ++j) {
+        EXPECT_FLOAT_EQ(g[j], g0[j]);
+      }
+    }
+  }
+
+  // test a single function of expA*B
+  var f = sum(res_vv);
+  var f0 = sum(expAB);
+  stan::math::set_zero_all_adjoints();
+  f.grad(Avec, g);
+  stan::math::set_zero_all_adjoints();
+  f0.grad(Avec, g0);
+  for (size_t j = 0; j < g.size(); ++j) {
+    EXPECT_FLOAT_EQ(g[j], g0[j]);
+  }
+}
+
+TEST(MathMatrix, matrix_exp_action_vv) {
+  test_matrix_exp_action_vv<1, 1>();
+  test_matrix_exp_action_vv<1, 5>();
+  test_matrix_exp_action_vv<5, 1>();
+  test_matrix_exp_action_vd<5, 5>();
+}
+
+// TEST(MathMatrix, matrix_exp_action_clock) {
```

Comment:
This is great! Gives us the opportunity to reproduce the time test --- and of course, it should be commented out, since we de not want to run it every time we run the unit tests.